### PR TITLE
[plugin] split RH, add preferences, add version

### PR DESCRIPTION
**Pull request by grqz**
split RH, add preferences, add version field for yt-dlp-get-pot v0.1.0

### CODE REVIEW

File: plugin/yt_dlp_plugins/extractor/getpot_bgutil_http.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+        if error_msg := response_json.get('error'):
+            raise RequestError(error_msg)
+        if 'po_token' not in response_json:
+            raise RequestError('Server did not respond with a po_token')
+            # Should we return None here?
+
+        return response_json['po_token']
+
+@register_preference(BgUtilHTTPPotProviderRH)
+def bgutil_HTTP_getpot_peference(rh, request):
```
Comment (Brainicism) [inline plugin/yt_dlp_plugins/extractor/getpot_bgutil_http.py:55]: nit: `preference` is spelt wrong here

File: plugin/yt_dlp_plugins/extractor/getpot_bgutil_script.py [hunk 2]
```diff
@@ -102,4 +70,8 @@ def _get_pot_via_script(self, script_path, visitor_data, data_sync_id):
             return json.loads(script_data_resp)['poToken']
         except (json.JSONDecodeError, TypeError, KeyError) as e:
             raise RequestError(
-                f'Error parsing JSON response from _get_pot_via_script(caused by {str(e)})')
+                f'Error parsing JSON response from _get_pot_via_script (caused by {str(e)})')
+
+@register_preference(BgUtilScriptPotProviderRH)
+def bgutil_script_getpot_peference(rh, request):
```
Comment (Brainicism) [inline plugin/yt_dlp_plugins/extractor/getpot_bgutil_script.py:76]: nit: same here

File: plugin/yt_dlp_plugins/extractor/getpot_bgutil_http.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+            response_json = json.load(response)
+        except Exception as e:
+            raise RequestError(
+                f'Error parsing response JSON (caused by {str(e)}). response = {response.read().decode()}')
+
+        if error_msg := response_json.get('error'):
+            raise RequestError(error_msg)
+        if 'po_token' not in response_json:
+            raise RequestError('Server did not respond with a po_token')
+            # Should we return None here?
```
Comment (Brainicism) [inline plugin/yt_dlp_plugins/extractor/getpot_bgutil_http.py:50]: raising an exception already exits from the function, nothing can be returned afterwards

File: plugin/yt_dlp_plugins/extractor/getpot_bgutil_script.py [hunk 4]
```diff
@@ -4,81 +4,49 @@
...
-    _PROVIDER_NAME = 'BgUtilPot'
+class BgUtilScriptPotProviderRH(GetPOTProvider):
+    _PROVIDER_NAME = 'BgUtilScriptPot'
     _SUPPORTED_CLIENTS = ('web_creator', 'web', 'web_embedded', 'web_music')
+    VERSION = __version__
 
     def _validate_get_pot(self, client: str, ydl: YoutubeDL, visitor_data=None, data_sync_id=None, player_url=None, **kwargs):
+        script_path = ydl.get_info_extractor('Youtube')._configuration_arg(
+            'getpot_bgutil_script', [None], casesense=True)[0]
+        # TODO: Add default script path
```
Comment (Brainicism) [inline plugin/yt_dlp_plugins/extractor/getpot_bgutil_script.py:24]: We can leave this for a seperate PR if u want

**Review by Brainicism — CHANGES_REQUESTED** [review]

File: plugin/yt_dlp_plugins/extractor/getpot_bgutil_http.py (hunk 1 above)
Comment (grqz) [reply plugin/yt_dlp_plugins/extractor/getpot_bgutil_http.py:53]:
```suggestion
def bgutil_HTTP_getpot_preference(rh, request):
```

File: plugin/yt_dlp_plugins/extractor/getpot_bgutil_script.py (hunk 2 above)
Comment (grqz) [reply plugin/yt_dlp_plugins/extractor/getpot_bgutil_script.py:74]:
```suggestion
def bgutil_script_getpot_preference(rh, request):
```

File: plugin/yt_dlp_plugins/extractor/getpot_bgutil_http.py (hunk 3 above)
Comment (grqz) [reply plugin/yt_dlp_plugins/extractor/getpot_bgutil_http.py:48]: i meant should we return None instead of raising an exception?

Comment (Brainicism) [reply plugin/yt_dlp_plugins/extractor/getpot_bgutil_http.py:48]: Raising the exception will let the PO framework fallback to another RH if needed right?

File: plugin/yt_dlp_plugins/extractor/getpot_bgutil_http.py [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    _SUPPORTED_CLIENTS = ('web_creator', 'web', 'web_embedded', 'web_music')
+    VERSION = __version__
+
+    def _validate_get_pot(self, client: str, ydl: YoutubeDL, visitor_data=None, data_sync_id=None, player_url=None, **kwargs):
+        base_url = ydl.get_info_extractor('Youtube')._configuration_arg(
+            'getpot_bgutil_baseurl', ['http://127.0.0.1:4416'], casesense=True)[0]
+        if not data_sync_id and not visitor_data:
+            raise UnsupportedRequest(
+                'One of [data_sync_id, visitor_data] must be passed')
+        # TODO: Ping the server
```
Comment (grqz) [inline plugin/yt_dlp_plugins/extractor/getpot_bgutil_http.py:22]: maybe do a separate pr

File: plugin/yt_dlp_plugins/extractor/getpot_bgutil_http.py (hunk 3 above)
Comment (grqz) [reply plugin/yt_dlp_plugins/extractor/getpot_bgutil_http.py:48]:
```suggestion
```

**Review by Brainicism — APPROVED** [review]

File: plugin/yt_dlp_plugins/extractor/getpot_bgutil_script.py (hunk 4 above)
Comment (grqz) [reply plugin/yt_dlp_plugins/extractor/getpot_bgutil_script.py]: done in 4b5d6ef13607e34519e05584a553685d65b5e938

File: plugin/yt_dlp_plugins/extractor/getpot_bgutil_script.py [hunk 6]
```diff
@@ -16,10 +16,15 @@ class BgUtilScriptPotProviderRH(GetPOTProvider):
     _SUPPORTED_CLIENTS = ('web_creator', 'web', 'web_embedded', 'web_music')
     VERSION = __version__
 
+    @classproperty(cache=True)
+    def _default_script_path(self):
+        home = os.path.expanduser("~")
+        return os.path.join(
+            home, 'bgutil-ytdlp-pot-provider', 'server', 'build', 'generate_once.js')
```
[the remark this answers is not in the document]
Comment (Brainicism) [reply plugin/yt_dlp_plugins/extractor/getpot_bgutil_script.py:23]: this is fine

**Comment by Brainicism** [comment]
TODO: README update to give context on how we decide which RH to use, what happens if both RHs are usable.

File: README.md [hunk 7]
```diff
@@ -60,20 +60,22 @@ node build/main.js
...
+1. Transpile the generation script to Javascript:
 
 ```shell
 cd server/
 yarn install --frozen-lockfile
 npx tsc
 ```
 
-Make sure `node` is available in your `PATH`.
+2. Move the `server` directory with transpiled Javascript to `~/bgutil-ytdlp-pot-provider/` if you're using Linux or MacOS.
```
Comment (Brainicism) [inline README.md:76]: Would it be easier to understand if we just said to download the source code to their home directory? Instead of asking them to move a specific directory into a new folder?

Comment (grqz) [reply README.md:76]: maybe add a `clone the repo/download the source code` step

File: README.md [hunk 8]
```diff
@@ -60,20 +61,24 @@ node build/main.js
...
-- **POST /invalidate_caches**: Resets the PO token cache, forcing new tokens to be generated on next fetch
+- **POST /invalidate_caches**: Resets the PO token cache, forcing new tokens to be generated on next fetch.
 </details>
 
 #### (b) Generation Script Option
 
-The generation script needs to be transpiled to Javascript before it can be used by the plugin.
+1. Download the latest source code (.zip/.tar.gz) from https://github.com/Brainicism/bgutil-ytdlp-pot-provider/releases. Extract the contents into your home directory (`~/` on Unix-based systems, `%USERPROFILE%` for Windows) if you want to use this method without needing to specify `getpot_bgutil_script` extractor argument on each yt-dlp invocation.
+
+2. Rename the folder to `bgutils-ytdlp-pot-provider`
```
Comment (grqz) [inline README.md:71]: the same here
